UnrealPlugin/TestingProject/Source/SlateUGS/Private/Tests/GsdkTests.cpp

```diff
@@ -0,0 +1,382 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#if (WITH_DEV_AUTOMATION_TESTS && WITH_EDITOR)
+#include "GsdkTests.h"
+#include "PlayFabGSDK.h"
+#include "GSDKUtils.h"
+#include "TestConfiguration.h"
+
+BEGIN_DEFINE_SPEC(AutomationSpec, "-GSDK", EAutomationTestFlags::EditorContext | EAutomationTestFlags::EngineFilter)
+const FString ConfigFilePath = FPaths::Combine(FPaths::ProjectDir(), TEXT("/Tests/testConfig.json"));
+const FString heartbeatEndpoint = "testEndpoint";
+const FString serverId = "testServerId";
+const FString logFolder = "testLogFolder";
+const FString sharedContentFolder = "testSharedContentFolder";
+const FString certFolder = "testCertFolder";
+const FString titleId = "testTitleId";
+const FString buildId = "testBuildId";
+const FString region = "testRegion";
+TestConfiguration config;
+FEventRef readyForPlayers{ EEventMode::ManualReset };
+FEventRef shutdownCalled{ EEventMode::ManualReset };
+END_DEFINE_SPEC(AutomationSpec)
+
+void AutomationSpec::Define()
+{
+    Describe("Tests", [this]()
+        {
+            BeforeEach([this]()
+                {
+                    FPlatformMisc::SetEnvironmentVar(TEXT("PF_TITLE_ID"), TEXT("testTitleId"));
+                    FPlatformMisc::SetEnvironmentVar(TEXT("PF_BUILD_ID"), TEXT("testBuildId"));
+                    FPlatformMisc::SetEnvironmentVar(TEXT("PF_REGION"), TEXT("testRegion"));
+                    FPlatformMisc::SetEnvironmentVar(TEXT("GSDK_CONFIG_FILE"), *ConfigFilePath);
+                    
+                    config = TestConfiguration();
+                });
+
+            AfterEach([this]()
+                {
+                    FPlayFabGSDKModule::Get().ResetInternalState();
+
+                    if (FPaths::FileExists(ConfigFilePath))
+                    {
+                        FPlatformFileManager::Get().GetPlatformFile().DeleteFile(*ConfigFilePath);
+                    }
+                });
+
+            It("ConfigAllSetInitializesFine", [this]()
+                {
+                    config.SetHeartbeatEndpoint(heartbeatEndpoint);
```

Comment:
General comment for all those tests. I am not quite following how the states of GSDK change / how - where is the test controller?

Could you add a ReadMe? This is great work so it would be nice to have it documented!

Also, I assume all these tests are functional and run at 100% success rate, right?

Reply:
On that note, asking a naive question - any chance these tests can run on GitHub action when a new PR is created?

Reply:
Yes this will be part of the PR gate that's coming later.

There is no defined overall test controller, just the setup/cleanup steps defined in this "AutomationSpec" to ensure each test is entirely self-contained. I can put a readme task on the backlog. Overall the tests exactly mirror the cpp/Unity tests.